libp2p/protocols/pubsub/pubsub.nim

```diff
@@ -47,61 +48,74 @@ type
-    observers: ref seq[PubSubObserver] # ref as in smart_ptr
-    msgIdProvider*: MsgIdProvider      # Turn message into message id (not nil)
+    observers: ref seq[PubSubObserver]              # ref as in smart_ptr
+    msgIdProvider*: MsgIdProvider                   # Turn message into message id (not nil)
     msgSeqno*: uint64
+    lifetimeFut*: Future[void]                      # pubsub liftime future
 
-method handleDisconnect*(p: PubSub, peer: PubSubPeer) {.base.} =
+method unsubscribePeer*(p: PubSub, peerId: PeerID) {.base.} =
   ## handle peer disconnects
   ##
-  
-  if not(isNil(peer)) and peer.peerInfo notin p.conns:
-    trace "deleting peer", peer = peer.id
-    peer.onConnect.fire() # Make sure all pending sends are unblocked
-    p.peers.del(peer.id)
-    trace "peer disconnected", peer = peer.id
 
-    # metrics
-    libp2p_pubsub_peers.set(p.peers.len.int64)
+  trace "unsubscribing pubsub peer", peer = $peerId
+  if peerId in p.peers:
+    p.peers.del(peerId)
 
-proc onConnClose(p: PubSub, conn: Connection) {.async.} =
-  try:
-    let peer = conn.peerInfo
-    await conn.closeEvent.wait()
+  libp2p_pubsub_peers.set(p.peers.len.int64)
 
-    if peer in p.conns:
-      p.conns[peer].excl(conn)
-      if p.conns[peer].len <= 0:
-        p.conns.del(peer)
+proc broadcast*(p: PubSub,
+                sendPeers: HashSet[PubSubPeer] | seq[PubSubPeer] | array[1, PubSubPeer],
+                msg: RPCMsg,
+                timeout: Duration): Future[int] {.async.} =
+  # send messages and cleanup failed peers
+  var sent: seq[tuple[id: PeerID, fut: Future[void]]]
+  for sendPeer in sendPeers:
+    if sendPeer.isNil:
+      continue
 
-    if peer.id in p.peers:
-      p.handleDisconnect(p.peers[peer.id])
+    # avoid sending to self
+    if sendPeer.peerId == p.peerInfo.peerId:
+      continue
 
-  except CancelledError as exc:
-    raise exc
-  except CatchableError as exc:
-    trace "exception in onConnClose handler", exc = exc.msg
+    trace "broadcast to peer", peer = sendPeer.id, message = msg
+    sent.add((id: sendPeer.peerId, fut: sendPeer.send(msg, timeout)))
+
+  var broadcasted: seq[PeerID]
+  var failed: seq[PeerID]
+  let futs = await allFinished(sent.mapIt(it.fut))
+  for s in futs:
+    let f = sent.filterIt(it.fut == s)
+    if f.len > 0:
+      if s.failed:
+        trace "broadcast to peer failed", peer = f[0].id
+        p.unsubscribePeer(f[0].id)
+        failed.add(f[0].id)
+      else:
+        trace "broadcast to peer succeeded", peer = f[0].id
+        broadcasted.add(f[0].id)
+
+  return broadcasted.len
 
 proc sendSubs*(p: PubSub,
                peer: PubSubPeer,
                topics: seq[string],
                subscribe: bool) {.async.} =
   ## send subscriptions to remote peer
-  asyncCheck peer.sendSubOpts(topics, subscribe)
+  discard await p.broadcast([peer], RPCMsg(subscriptions: topics.mapIt(SubOpts(subscribe: subscribe, topic: it))), DefaultSendTimeout)
```

Comment:
using broadcast here looks like a long and winding road compared to simply doing `peer.send`..?

Reply:
ie the `use-only-broadcast` idea may be fine, but this takes it over the top - semantically, we're not broadcasting here, and the extra future traffic is weird - in addition, keep in mind that the generic code above creates a fair amount of bloat - ie the compiler will generate separate/duplicate  C functions for every type in the signature multiplied by all the continuations, exception bloat etc

Reply:
We use broadcast to unify the maintenance to do in the case the send fails. This was discussed with @dryajov on discord. 
Basically a pure `send` right now might not perform the required maintenance.

Maybe broadcast is not the right name btw.

Btw I'm a bit tired of hearing that nim X or Y feature should not be used. Generics are a good thing. Much easier to understand then a macro or template...

Reply:
But I just realized this was an asyncCheck! That might be better stay. ( maybe that's what you meant )

Reply:
sounds like a pure send that maintains things is needed then - ie it's always easier to reason about _one_ thing that sends, deal with the tricky issues, and then accumulate the results - trying to do both will often lead to issues when errors happen in the middle of the loop for example

Reply:
Well, that's what `broadcast` is for, it deals with sending messages to peers and updating the state of the mesh according to the outcome. `PubSubPeer`'s `send` on the other hand, deals with maintaining state for one peer.

Reply:
my point is more that there's no reason for the unsubscribepeer to be in a proc that _also_ deals with accumulating a result - the cleanup is done per-peer independently, so following the logic in a single-peer-send will be more easy to reason about because it will not be interspersed with peer looping logic

Reply:
Yes, but it's not possible because the peer knows nothing about the mesh state, that's all dealt by the specific pubsub implementation, ie the peer would not know which table to remove itself from. I would also much prefered this was dealt right at the point of send in the peer, but it isn't possible and this is the next best thing.

Reply:
this is a very common source of bugs in the libp2p code that is trivial to avoid by simply placing the logic at the abstraction level where it belongs - if a peer send fails, it's independent of other peer sends failing, in gossipsub - therefore, it belongs somewhere where only a single peer is being dealt with - the code will be much easier to read and debug if it follows this simple principle.


consider for a minute what you're doing: you're initiating a bunch of sends in a loop where you have a reference to the peer. you then save the id, forget about this peer reference, and start a new loop to do the cleanup, doing the lookup of the peer by id _again_ - not only does this introduce the potential for race conditions - it is also inefficient and harder to reason about.

Reply:
> not possible

`proc send(p: PubSub, peer: PubSubPeer, msg:seq[byte]) {.async.}` ?

Reply:
@sinkingsugar @arnetheduck has a point, we can further simplify it by moving it to a pubsub `.send(...)` method.

Reply:
@dryajov you took care of it I see, nothing else to add on my side right?

Reply:
We should be good for now.